### ISSUE TITLE
Changes to `rust-version` field in `Cargo.toml` that change Clippy's MSRV detection do not bust the compilation cache

### ISSUE THREAD

**Issue by lopopolo** [rust-lang/rust-clippy#10105]
### Summary

Running `cargo clippy --workspace --all-targets --all-features` on code with a `clippy::uninlined_format_args` lint violation followed by setting `rust-version = "1.52.0"` in `Cargo.toml` results in Clippy still reporting the `uninlined_format_args` lint violation on subsequent runs of `cargo clippy --workspace --all-targets --all-features`.

This is only able to be resolved with a `cargo clean`.

### Reproducer

I tried this code:

```rust
<code>
```

I expected to see this happen:

Instead, this happened:

### Version

```text
rustc 1.66.0 (69f9c33d7 2022-12-12)
binary: rustc
commit-hash: 69f9c33d71c871fc16ac445211281c6e7a340943
commit-date: 2022-12-12
host: x86_64-apple-darwin
release: 1.66.0
LLVM version: 15.0.2
```

**Comment by Alexendoo** [comment]
Will be fixed by #9707 in 1.67
